main.tf

```
@@ -19,7 +19,11 @@ locals {
   folders_list = [for name in var.names : google_folder.folders[name]]
   first_folder = local.folders_list[0]
 
-  name_role_pairs = setproduct(var.names, var.folder_admin_roles)
+  name_role_pairs = concat(
+    [for name in var.names : setproduct(
+      [name], lookup(var.per_folder_admins, name, null).roles != null ? lookup(var.per_folder_admins, name, null).roles : var.folder_admin_roles
```

nit: since we validate ` lookup(var.per_folder_admins, name, null).roles != null`, I think we can directly access in true case.
```suggestion
      [name], lookup(var.per_folder_admins, name, null).roles != null ? var.per_folder_admins[name].roles : var.folder_admin_roles
```

---

Also wondering if the logic will work when we have elements in var.names that are not keys in var.per_folder_admins since `lookup(var.per_folder_admins, name, null)` will eval to `null` and then would null.roles be valid?

---

I've simplified the logic now.